### PR TITLE
Get started with a Lexical Analyzer

### LINKED ISSUE #4931: Get started with a Lexical Analyzer
#### Proposed title of article

Get started with a Lexical Analyzer.

#### Proposed article introduction

[Languages] In the older computer generations, running simple codes were much trickier than thought. This is because everything was done in Machine languages written on punch cards. It then went to mnemonics as used in the Assembly language. These were simpler but still required much understanding of machine language.
The need for high-level languages was higher than ever since computers seem effective in repetitive tasks.

The invention of a high-level language compiler was a significant breakthrough in the computer growth phase. This breakthrough was because high-level languages are understandable by people though much overhead is needed. Hence, for one to be able to appreciate the compilers and their work in code conversion from source to optimized object code, one has to understand how they are designed.

Compiler design involves several phases. Such phases include the following:

- Lexical analysis
- Syntax analysis
- Semantic analysis
- intermediate code generation
- Code optimization
- Target code generation
- Error handling
- Symbol table generation

A Lexical analyzer is found in the first phase of compiler construction.

#### Key takeaways

In this article, one will learn the following:

- What is Lexical analysis
- Lexical analysis core principles
- Importance of a lexical analysis
- Application of the principles in creating a lexical analyzer
- Building a lexical analyzer with code examples

#### Article quality

The article focuses on the core principles used in the creation of a lexical analyzer. It then speaks of how to use the principles in creating a lexical analyzer.
Running a lexical analyzer with code samples is also among the key points to be discussed.

#### References

- Lexical analysis [definitions](https://en.wikipedia.org/wiki/Lexical_analysis)
- Lexical analysis research [document](https://webcache.googleusercontent.com/search?q=cache:XJzw8xH2YtgJ:https://www.iith.ac.in/~ramakrishna/Compilers-Aug14/doc/flex.pdf+&cd=9&hl=en&ct=clnk&gl=ke)
- Lexical analyzer [IBM documentation](https://www.ibm.com/docs/en/zos/2.2.0?topic=section-lexical-analyzer)

### Templates to use as guides

- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by franciskaguongo**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

#### Topic suggestion this closes
This closes #4931

### CODE REVIEW

**Comment by franciskaguongo** [comment]
@srishilesh I think the article has a hero image

File: content/articles/get-started-with-a-lexical-analyzer/index.md [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+Lex files have a `.l` extension.
+However, before creating a Lex file, let us first know the parts of a Lex file.
+
+#### Lex file parts
+Lex files have three main parts. These parts are the following:
+- Definition
+- Rules
+- Subroutines
+
+**Definition**: It contains literal blocks, definitions, internal table declarations, start conditions, and translations.
```
Comment (srishilesh) [inline content/articles/get-started-with-a-lexical-analyzer/index.md:182]:
```suggestion
##### Definition
It contains literal blocks, definitions, internal table declarations, start conditions, and translations.
```

File: content/articles/get-started-with-a-lexical-analyzer/index.md [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+It can contain things like the 'include' section found in the C program. This statement states the things that the program could need.
+Definitions can be written inside the `%{` and `%}` signs.
+
+**Rules and actions**: This section carries patterns and _C_ code(actions). These patterns can be in ECMAScript RegEx or wildcard format. It specifies the tokens the program checks, and the C code states what happens when they are found in the Actions section.
```
Comment (srishilesh) [inline content/articles/get-started-with-a-lexical-analyzer/index.md:186]:
```suggestion
##### Rules and actions
This section carries patterns and _C_ code(actions). These patterns can be in ECMAScript RegEx or wildcard format. It specifies the tokens the program checks, and the C code states what happens when they are found in the Actions section.
```

File: content/articles/get-started-with-a-lexical-analyzer/index.md [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+Actions can be found inside the curly brackets({ ... }).
+The actions or macros can be optional too. Hence, they can be left blank.
+
+**Subroutines**: C codes that state additional user-defined functions executed when the patterns are found or not found.
```
Comment (srishilesh) [inline content/articles/get-started-with-a-lexical-analyzer/index.md:190]:
```suggestion
##### Subroutines
C codes that state additional user-defined functions executed when the patterns are found or not found.
```

File: content/articles/get-started-with-a-lexical-analyzer/index.md [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+2. Rules section:
+
+regular expression        { c code to be executed when this R.E. is matched against the source. }
+another regular expression { another c code based on this RegEX }
+%%
+
+3. User Sub-routines:
+
+Additional code here
+```
```
Comment (srishilesh) [inline content/articles/get-started-with-a-lexical-analyzer/index.md:215]: Why are these enclosed within ` `` `?

If these are not code snippets, they need not be enclosed

File: content/articles/get-started-with-a-lexical-analyzer/index.md [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+1. Definitions section:
+
+%{    contains C code, usually consisting of the libraries
+        #define external global declarations that one wants to be part of the code for y.tab.c
+%}
+
+The file can include optional macro definitions for regular expressions.
+
+%%
```
Comment (srishilesh) [inline content/articles/get-started-with-a-lexical-analyzer/index.md:204]: This is not very clear. Please explain more about this.

File: content/articles/get-started-with-a-lexical-analyzer/index.md [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+
+**Rules and actions**: This section carries patterns and _C_ code(actions). These patterns can be in ECMAScript RegEx or wildcard format. It specifies the tokens the program checks, and the C code states what happens when they are found in the Actions section.
+Actions can be found inside the curly brackets({ ... }).
+The actions or macros can be optional too. Hence, they can be left blank.
+
+**Subroutines**: C codes that state additional user-defined functions executed when the patterns are found or not found.
+They can also contain other functions that the program may execute on the run.
+These are optional.
+
+The three sections are divided by delimiters (`%%`) as shown below:
```
Comment (srishilesh) [inline content/articles/get-started-with-a-lexical-analyzer/index.md:194]: Are these part of code?

File: content/articles/get-started-with-a-lexical-analyzer/index.md [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+/* Prints the output */
+printf("This file contains the following:");
+printf("\n\t%d lines", lines);
+printf("\n\t%d words", words);
+printf("\n\t%d Lowercase letters", lowercase_letters);
+printf("\n\t%d Uppercase letters", uppercase_letters);
+printf("\n\t%d digits", numbers);
+printf("\n\t%d special characters", special_characters);
+printf("\n\t The total size of the file characters in bytes is: %d bytes.\n", size);
+}
```
Comment (srishilesh) [inline content/articles/get-started-with-a-lexical-analyzer/index.md:271]: Can you format these?

File: content/articles/get-started-with-a-lexical-analyzer/index.md [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+
+printf("%s",yytext);
+
+}
+
+.* {ECHO;}
+
+\n {printf("%s",yytext);}
+
```
Comment (srishilesh) [inline content/articles/get-started-with-a-lexical-analyzer/index.md:378]: Format these

**Review by srishilesh — CHANGES_REQUESTED** [review]
@franciskaguongo 

Good to see you again.

Overall markdown formatting looks good, but the indentation of code snippets doesn't look good.

Please go through the suggestions I've made.

Ping me when done.

Also, let's wait for the plagiarism check before we start the tech review.

Thanks.

File: content/articles/get-started-with-a-lexical-analyzer/index.md (hunk 1 above)
Comment (franciskaguongo) [reply content/articles/get-started-with-a-lexical-analyzer/index.md]: ok, got you

File: content/articles/get-started-with-a-lexical-analyzer/index.md (hunk 4 above)
Comment (franciskaguongo) [reply content/articles/get-started-with-a-lexical-analyzer/index.md:218]: They are code snippets showing the structure of the code in the lexical file

**Comment by franciskaguongo** [comment]
> Definitions section:

nice to see you again

**Comment by franciskaguongo** [comment]
> @franciskaguongo
> 
> Good to see you again.
> 
> Overall markdown formatting looks good, but the indentation of code snippets doesn't look good.
> 
> Please go through the suggestions I've made.
> 
> Ping me when done.
> 
> Also, let's wait for the plagiarism check before we start the tech review.
> 
> Thanks.

Nice to see you again

File: content/articles/get-started-with-a-lexical-analyzer/index.md (hunk 6 above)
Comment (franciskaguongo) [reply content/articles/get-started-with-a-lexical-analyzer/index.md:197]: no, but explain how the code looks like

File: content/articles/get-started-with-a-lexical-analyzer/index.md (hunk 7 above)
Comment (franciskaguongo) [reply content/articles/get-started-with-a-lexical-analyzer/index.md:274]: when you add indentation on lexical files, the whole app breaks. They are very sensitive to structure and indentation

File: content/articles/get-started-with-a-lexical-analyzer/index.md (hunk 8 above)
Comment (franciskaguongo) [reply content/articles/get-started-with-a-lexical-analyzer/index.md:381]: It's quite hard; when you add indentation on lexical files, the whole app breaks. They are very sensitive to structure and indentation

File: content/articles/get-started-with-a-lexical-analyzer/index.md (hunk 4 above)
Comment (srishilesh) [reply content/articles/get-started-with-a-lexical-analyzer/index.md]: @franciskaguongo

You mean `1. Definitions section:` and `2. Rules section:` are part of code snippets?

Can you add screenshots here instead? It looks misleading.

File: content/articles/get-started-with-a-lexical-analyzer/index.md (hunk 7 above)
Comment (srishilesh) [reply content/articles/get-started-with-a-lexical-analyzer/index.md:277]: Got it. Thanks for clarification.

**Comment by srishilesh** [comment]
@franciskaguongo 

The current changes looks great.

I've replied for a few, please go through them, and make necessary changes.

Ping me when you're done.

**Comment by hectorkambow** [comment]
@srishilesh  

It may be best to lay out anything else that needed to be fixed to avoid multiple back and forth.

**Comment by hectorkambow** [comment]
Jumping on this for a plagiarism check now

**Comment by hectorkambow** [comment]
Plagiarism check looks good. 👀  

cc @section-engineering-education/content-moderator-team  

[New Scan 5_32 PM Copyleaks Report.pdf](https://github.com/section-engineering-education/engineering-education/files/7636075/New.Scan.5_32.PM.Copyleaks.Report.pdf)

File: content/articles/get-started-with-a-lexical-analyzer/index.md (hunk 5 above)
Comment (franciskaguongo) [reply content/articles/get-started-with-a-lexical-analyzer/index.md]: let me check into it

File: content/articles/get-started-with-a-lexical-analyzer/index.md (hunk 4 above)
Comment (franciskaguongo) [reply content/articles/get-started-with-a-lexical-analyzer/index.md]: let me look into it

**Comment by srishilesh** [comment]
@franciskaguongo 

Let me know once you're done the suggestions.

Thanks.

File: content/articles/get-started-with-a-lexical-analyzer/index.md (hunk 4 above)
Comment (franciskaguongo) [reply content/articles/get-started-with-a-lexical-analyzer/index.md]: Done

File: content/articles/get-started-with-a-lexical-analyzer/index.md (hunk 5 above)
Comment (franciskaguongo) [reply content/articles/get-started-with-a-lexical-analyzer/index.md]: @srishilesh, these were previously spoken of in the parts before. The parts under the definition, rules, and subroutines headers in the article. Once the reader reads them (they are a little bit in-depth) when they reach this part, they will remember the previously read through brief hints in the screenshot

Comment (franciskaguongo) [reply content/articles/get-started-with-a-lexical-analyzer/index.md]: Please review and tell me if it is well

**Comment by franciskaguongo** [comment]
> @franciskaguongo
> 
> Let me know once you're done the suggestions.
> 
> Thanks.

Done... :+1:

**Comment by franciskaguongo** [comment]
@srishilesh, I will be waiting for your ping once the tech review is done! Thanks...

**Comment by hectorkambow** [comment]
@srishilesh 

Just looking to see if the PR is still in a tech review - if the articles needs to be polished up some more. 

We will be closing the PR 👍 to allow the author to make the edits out of the queue.

**Comment by srishilesh** [comment]
> @srishilesh 
> 
> Just looking to see if the PR is still in a tech review - if the articles needs to be polished up some more. 
> 
> We will be closing the PR 👍 to allow the author to make the edits out of the queue.

I'll complete my tech review by EOD.

**Review by srishilesh — APPROVED** [review]
@franciskaguongo 

Yet another amazing article.

Apologies to keep you waiting, in spite of you pinging me via Slack several times. I'll try to provide faster reviews from the next time.

I've made minor changes to flow and formatting. Make sure to go through them.

Moving for the final review.

Hoping to work with you soon.

Thanks.

**Comment by LinusMuema** [comment]
Hi @franciskaguongo 

Awesome job on the article.

cc: @srishilesh

**Pull request merged**